File: tdz/src/tdz.rs
use crate::error::TodoziError;
use regex::Regex;
use serde_json::Value;
use std::collections::HashMap;
use std::env;
#[derive(Debug, Clone)]
pub struct TdzCommand {
    pub command: String,
    pub target: String,
    pub parameters: Vec<String>,
    pub options: HashMap<String, String>,
}
pub fn find_todozi(str: Option<&str>) -> Option<String> {
    let home = env::var("HOME").ok()?;
    let base = format!("{}/.todozi", home);
    if str.is_some() { Some(format!("{}/{}", base, str.unwrap())) } else { Some(base) }
}
pub fn parse_tdz_command(text: &str) -> Result<Vec<TdzCommand>, TodoziError> {
    let mut commands = Vec::new();
    let re = Regex::new(r"<tdz>(.*?)</tdz>")
        .map_err(|e| TodoziError::ValidationError {
            message: e.to_string(),
        })?;
    for cap in re.captures_iter(text) {
        let content = cap[1].trim();
        let parts: Vec<&str> = content.split(';').map(|s| s.trim()).collect();
        if parts.is_empty() {
            continue;
        }
        let command = parts[0].to_lowercase();
        let target = parts.get(1).map(|s| s.to_lowercase()).unwrap_or_default();
        let mut parameters = Vec::new();
        let mut options = HashMap::new();
        for part in &parts[2..] {
            if part.contains('=') {
                let kv: Vec<&str> = part.splitn(2, '=').collect();
                if kv.len() == 2 {
                    options.insert(kv[0].to_lowercase(), kv[1].to_string());
                }
            } else {
                parameters.push(part.to_string());
            }
        }
        commands
            .push(TdzCommand {
                command,
                target,
                parameters,
                options,
            });
    }
    Ok(commands)
}
pub async fn execute_tdz_command(
    command: &TdzCommand,
    base_url: &str,
    api_key: Option<&str>,
) -> Result<Value, TodoziError> {
    use reqwest::Client;
    let client = Client::new();
    let url = format!(
        "{}{}", base_url.trim_end_matches('/'), get_endpoint_path(command)
    );
    let mut request = match command.command.as_str() {
        "list" | "get" | "search" => client.get(&url),
        "create" => client.post(&url),
        "update" => client.put(&url),
        "delete" => client.delete(&url),
        "run" | "execute" => client.post(&url),
        _ => {
            return Err(TodoziError::ValidationError {
                message: format!("Unknown command: {}", command.command),
            });
        }
    };
    if let Some(key) = api_key {
        request = request.header("X-API-Key", key);
    }
    match command.command.as_str() {
        "create" | "update" => {
            let body = build_request_body(command)?;
            request = request.json(&body);
        }
        "run" | "execute" => {
            let body = build_run_body(command)?;
            request = request.json(&body);
        }
        _ => {}
    }
    let response = request
        .send()
        .await
        .map_err(|e| TodoziError::ValidationError {
            message: format!("Network error: {}", e),
        })?;
    let status = response.status();
    if !status.is_success() {
        let error_text = response.text().await.unwrap_or_default();
        return Err(TodoziError::ValidationError {
            message: format!("HTTP error {}: {}", status, error_text),
        });
    }
    let result: Value = response
        .json()
        .await
        .map_err(|e| TodoziError::ValidationError {
            message: format!("JSON parse error: {}", e),
        })?;
    Ok(result)
}
fn get_endpoint_path(command: &TdzCommand) -> String {
    match (command.command.as_str(), command.target.as_str()) {
        ("list" | "get", "health") => "/health".to_string(),
        ("list" | "get", "stats") => "/stats".to_string(),
        ("run", "init") => "/init".to_string(),
        ("list", "tasks") => "/tasks".to_string(),
        ("get", "task") => {
            format!("/tasks/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("create", "task") => "/tasks".to_string(),
        ("update", "task") => {
            format!("/tasks/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("delete", "task") => {
            format!("/tasks/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("search", "tasks") => {
            format!(
                "/tasks/search?q={}", command.parameters.get(0).unwrap_or(& ""
                .to_string())
            )
        }
        ("list", "memories") => "/memories".to_string(),
        ("list", "memories_secret") => "/memories/secret".to_string(),
        ("list", "memories_human") => "/memories/human".to_string(),
        ("list", "memories_short") => "/memories/short".to_string(),
        ("list", "memories_long") => "/memories/long".to_string(),
        ("create", "memory") => "/memories".to_string(),
        ("list", "ideas") => "/ideas".to_string(),
        ("create", "idea") => "/ideas".to_string(),
        ("list", "agents") => "/agents".to_string(),
        ("list", "agents_available") => "/agents/available".to_string(),
        ("get", "agent") => {
            format!("/agents/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("get", "agent_status") => {
            format!(
                "/agents/{}/status", command.parameters.get(0).unwrap_or(& ""
                .to_string())
            )
        }
        ("create", "agent") => "/agents".to_string(),
        ("update", "agent") => {
            format!("/agents/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("delete", "agent") => {
            format!("/agents/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("run", "agent") => {
            format!(
                "/chat/agent/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("list", "training") => "/training".to_string(),
        ("get", "training") => {
            format!(
                "/training/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("create", "training") => "/training".to_string(),
        ("update", "training") => {
            format!(
                "/training/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("delete", "training") => {
            format!(
                "/training/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("run", "training_export") => "/training/export".to_string(),
        ("list", "training_stats") => "/training/stats".to_string(),
        ("run", "chat") => "/chat/process".to_string(),
        ("list", "chat_history") => "/chat/history".to_string(),
        ("list", "analytics_tasks") => "/analytics/tasks".to_string(),
        ("list", "analytics_agents") => "/analytics/agents".to_string(),
        ("list", "analytics_performance") => "/analytics/performance".to_string(),
        ("run", "time_start") => {
            format!(
                "/time/start/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("run", "time_stop") => {
            format!(
                "/time/stop/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("list", "time_report") => "/time/report".to_string(),
        ("list", "chunks") => "/chunks".to_string(),
        ("list", "chunks_ready") => "/chunks/ready".to_string(),
        ("list", "chunks_graph") => "/chunks/graph".to_string(),
        ("create", "chunk") => "/chunks".to_string(),
        ("list", "projects") => "/projects".to_string(),
        ("create", "project") => "/projects".to_string(),
        ("list", "feelings") => "/feelings".to_string(),
        ("get", "feeling") => {
            format!(
                "/feelings/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("create", "feeling") => "/feelings".to_string(),
        ("update", "feeling") => {
            format!(
                "/feelings/{}/{}", command.parameters.get(0).unwrap_or(& "".to_string()),
                command.parameters.get(1).unwrap_or(& "".to_string())
            )
        }
        ("delete", "feeling") => {
            format!(
                "/feelings/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("list", "errors") => "/errors".to_string(),
        ("get", "error") => {
            format!("/errors/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("create", "error") => "/errors".to_string(),
        ("update", "error") => {
            format!("/errors/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("delete", "error") => {
            format!("/errors/{}", command.parameters.get(0).unwrap_or(& "".to_string()))
        }
        ("search", "errors") => {
            format!(
                "/errors/search?q={}", command.parameters.get(0).unwrap_or(& ""
                .to_string())
            )
        }
        ("create", "queue_item") => "/queue/plan".to_string(),
        ("list", "queue") => "/queue/list".to_string(),
        ("list", "queue_backlog") => "/queue/list/backlog".to_string(),
        ("list", "queue_active") => "/queue/list/active".to_string(),
        ("list", "queue_complete") => "/queue/list/complete".to_string(),
        ("run", "queue_start") => {
            format!(
                "/queue/start/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("run", "queue_end") => {
            format!(
                "/queue/end/{}", command.parameters.get(0).unwrap_or(& "".to_string())
            )
        }
        ("run", "api_register") => "/api/register".to_string(),
        ("run", "api_check") => "/api/check".to_string(),
        _ => format!("/{}", command.target),
    }
}
fn build_request_body(command: &TdzCommand) -> Result<Value, TodoziError> {
    match command.target.as_str() {
        "task" => {
            Ok(
                serde_json::json!(
                    { "action" : command.options.get("action").cloned()
                    .unwrap_or_default(), "time" : command.options.get("time").cloned()
                    .unwrap_or_default(), "priority" : command.options.get("priority")
                    .cloned().unwrap_or_default(), "project" : command.options
                    .get("project").cloned().unwrap_or_default(), "status" : command
                    .options.get("status").cloned().unwrap_or_default(), "assignee" :
                    command.options.get("assignee").cloned(), "tags" : command.options
                    .get("tags").map(| t | t.split(',').map(| s | s.trim().to_string())
                    .collect::< Vec < _ >> ()) }
                ),
            )
        }
        "memory" => {
            Ok(
                serde_json::json!(
                    { "moment" : command.options.get("moment").cloned()
                    .unwrap_or_default(), "meaning" : command.options.get("meaning")
                    .cloned().unwrap_or_default(), "reason" : command.options
                    .get("reason").cloned().unwrap_or_default(), "importance" : command
                    .options.get("importance").cloned().unwrap_or_default(), "term" :
                    command.options.get("term").cloned().unwrap_or_default(),
                    "memory_type" : command.options.get("memory_type").cloned()
                    .unwrap_or_default(), "emotion" : command.options.get("emotion")
                    .cloned() }
                ),
            )
        }
        "idea" => {
            Ok(
                serde_json::json!(
                    { "idea" : command.options.get("idea").cloned().unwrap_or_default(),
                    "share" : command.options.get("share").cloned().unwrap_or_default(),
                    "importance" : command.options.get("importance").cloned()
                    .unwrap_or_default() }
                ),
            )
        }
        "agent" => {
            Ok(
                serde_json::json!(
                    { "name" : command.options.get("name").cloned().unwrap_or_default(),
                    "description" : command.options.get("description").cloned()
                    .unwrap_or_default(), "capabilities" : command.options
                    .get("capabilities").map(| c | c.split(',').map(| s | s.trim()
                    .to_string()).collect::< Vec < _ >> ()) }
                ),
            )
        }
        "chunk" => {
            Ok(
                serde_json::json!(
                    { "id" : command.options.get("id").cloned().unwrap_or_default(),
                    "level" : command.options.get("level").cloned().unwrap_or_default(),
                    "description" : command.options.get("description").cloned()
                    .unwrap_or_default(), "dependencies" : command.options
                    .get("dependencies").map(| d | d.split(',').map(| s | s.trim()
                    .to_string()).collect::< Vec < _ >> ()), "code" : command.options
                    .get("code").cloned().unwrap_or_default() }
                ),
            )
        }
        "project" => {
            Ok(
                serde_json::json!(
                    { "name" : command.options.get("name").cloned().unwrap_or_default(),
                    "description" : command.options.get("description").cloned()
                    .unwrap_or_default(), "status" : command.options.get("status")
                    .cloned().unwrap_or_default() }
                ),
            )
        }
        "feeling" => {
            Ok(
                serde_json::json!(
                    { "emotion" : command.options.get("emotion").cloned()
                    .unwrap_or_default(), "intensity" : command.options.get("intensity")
                    .and_then(| s | s.parse::< u8 > ().ok()).unwrap_or(5), "description"
                    : command.options.get("description").cloned().unwrap_or_default(),
                    "context" : command.options.get("context").cloned(), "tags" : command
                    .options.get("tags").map(| t | t.split(',').map(| s | s.trim()
                    .to_string()).collect::< Vec < _ >> ()) }
                ),
            )
        }
        "training" => {
            Ok(
                serde_json::json!(
                    { "data_type" : command.options.get("data_type").cloned()
                    .unwrap_or_default(), "prompt" : command.options.get("prompt")
                    .cloned().unwrap_or_default(), "completion" : command.options
                    .get("completion").cloned().unwrap_or_default(), "source" : command
                    .options.get("source").cloned().unwrap_or_default(), "context" :
                    command.options.get("context").cloned(), "tags" : command.options
                    .get("tags").map(| t | t.split(',').map(| s | s.trim().to_string())
                    .collect::< Vec < _ >> ()), "quality_score" : command.options
                    .get("quality_score").and_then(| s | s.parse::< f32 > ().ok()) }
                ),
            )
        }
        _ => Ok(Value::Null),
    }
}
fn build_run_body(command: &TdzCommand) -> Result<Value, TodoziError> {
    match command.target.as_str() {
        "agent" => {
            Ok(
                serde_json::json!(
                    { "message" : command.options.get("message").cloned()
                    .unwrap_or_default(), "context" : command.options.get("context")
                    .cloned() }
                ),
            )
        }
        "chat" => {
            Ok(
                serde_json::json!(
                    { "message" : command.options.get("message").cloned()
                    .unwrap_or_default(), "context" : command.options.get("context")
                    .cloned() }
                ),
            )
        }
        "queue_start" | "queue_end" => Ok(Value::Null),
        "api_check" => {
            Ok(
                serde_json::json!(
                    { "public_key" : command.options.get("public_key").cloned()
                    .unwrap_or_default(), "private_key" : command.options
                    .get("private_key").cloned() }
                ),
            )
        }
        _ => Ok(Value::Null),
    }
}
pub async fn process_tdz_commands(
    text: &str,
    base_url: &str,
    api_key: Option<&str>,
) -> Result<Vec<Value>, TodoziError> {
    let commands = parse_tdz_command(text)?;
    let mut results = Vec::new();
    for command in commands {
        let result = execute_tdz_command(&command, base_url, api_key).await?;
        results.push(result);
    }
    Ok(results)
}

File: tdz_js/ts/tdz.ts
export interface TdzCommand {
    command: string;
    target: string;
    parameters: string[];
    options: Map<string, string>;
}

export class TodoziError extends Error {
    constructor(public message: string) {
        super(message);
        this.name = 'TodoziError';
    }
}

export function findTodozi(str?: string): string | null {
    const home = process.env.HOME;
    if (!home) return null;
    const base = `${home}/.todozi`;
    if (str) {
        return `${base}/${str}`;
    }
    return base;
}

export function parseTdzCommand(text: string): TdzCommand[] {
    const commands: TdzCommand[] = [];
    const re = /<tdz>(.*?)<\/tdz>/gs;
    let match;
    
    while ((match = re.exec(text)) !== null) {
        const content = match[1].trim();
        const parts = content.split(';').map(s => s.trim()).filter(s => s.length > 0);
        
        if (parts.length === 0) continue;
        
        const command = parts[0].toLowerCase();
        const target = parts[1] ? parts[1].toLowerCase() : '';
        const parameters: string[] = [];
        const options = new Map<string, string>();
        
        for (const part of parts.slice(2)) {
            if (part.includes('=')) {
                const kv = part.split('=', 2);
                if (kv.length === 2) {
                    options.set(kv[0].toLowerCase(), kv[1]);
                }
            } else {
                parameters.push(part);
            }
        }
        
        commands.push({
            command,
            target,
            parameters,
            options
        });
    }
    
    return commands;
}

export async function executeTdzCommand(
    command: TdzCommand,
    baseUrl: string,
    apiKey?: string
): Promise<any> {
    const url = `${baseUrl.replace(/\/$/, '')}${getEndpointPath(command)}`;
    
    let requestInit: RequestInit = {
        method: 'GET'
    };
    
    switch (command.command) {
        case 'list':
        case 'get':
        case 'search':
            requestInit.method = 'GET';
            break;
        case 'create':
            requestInit.method = 'POST';
            break;
        case 'update':
            requestInit.method = 'PUT';
            break;
        case 'delete':
            requestInit.method = 'DELETE';
            break;
        case 'run':
        case 'execute':
            requestInit.method = 'POST';
            break;
        default:
            throw new TodoziError(`Unknown command: ${command.command}`);
    }
    
    const headers: Record<string, string> = {};
    if (apiKey) {
        headers['X-API-Key'] = apiKey;
    }
    
    if (command.command === 'create' || command.command === 'update' || command.command === 'run' || command.command === 'execute') {
        headers['Content-Type'] = 'application/json';
        const body = command.command === 'run' || command.command === 'execute' 
            ? buildRunBody(command)
            : buildRequestBody(command);
        requestInit.body = JSON.stringify(body);
    }
    
    requestInit.headers = headers;
    
    const response = await fetch(url, requestInit);
    
    if (!response.ok) {
        const errorText = await response.text().catch(() => '');
        throw new TodoziError(`HTTP error ${response.status}: ${errorText}`);
    }
    
    const result = await response.json().catch(() => {
        throw new TodoziError('JSON parse error: Invalid response format');
    });
    
    return result;
}

function getEndpointPath(command: TdzCommand): string {
    const param0 = command.parameters[0] || '';
    const param1 = command.parameters[1] || '';
    
    switch (command.command) {
        case 'list':
        case 'get':
            switch (command.target) {
                case 'health': return '/health';
                case 'stats': return '/stats';
                case 'tasks': return '/tasks';
                case 'task': return `/tasks/${param0}`;
                case 'memories': return '/memories';
                case 'memories_secret': return '/memories/secret';
                case 'memories_human': return '/memories/human';
                case 'memories_short': return '/memories/short';
                case 'memories_long': return '/memories/long';
                case 'ideas': return '/ideas';
                case 'agents': return '/agents';
                case 'agents_available': return '/agents/available';
                case 'agent': return `/agents/${param0}`;
                case 'agent_status': return `/agents/${param0}/status`;
                case 'training': return `/training/${param0}`;
                case 'training_stats': return '/training/stats';
                case 'chat_history': return '/chat/history';
                case 'analytics_tasks': return '/analytics/tasks';
                case 'analytics_agents': return '/analytics/agents';
                case 'analytics_performance': return '/analytics/performance';
                case 'time_report': return '/time/report';
                case 'chunks': return '/chunks';
                case 'chunks_ready': return '/chunks/ready';
                case 'chunks_graph': return '/chunks/graph';
                case 'projects': return '/projects';
                case 'feelings': return '/feelings';
                case 'feeling': return `/feelings/${param0}`;
                case 'errors': return '/errors';
                case 'error': return `/errors/${param0}`;
                case 'queue': return '/queue/list';
                case 'queue_backlog': return '/queue/list/backlog';
                case 'queue_active': return '/queue/list/active';
                case 'queue_complete': return '/queue/list/complete';
                default: return `/${command.target}`;
            }
        case 'create':
            switch (command.target) {
                case 'task': return '/tasks';
                case 'memory': return '/memories';
                case 'idea': return '/ideas';
                case 'agent': return '/agents';
                case 'chunk': return '/chunks';
                case 'project': return '/projects';
                case 'feeling': return '/feelings';
                case 'error': return '/errors';
                case 'training': return '/training';
                case 'queue_item': return '/queue/plan';
                default: return `/${command.target}`;
            }
        case 'update':
            switch (command.target) {
                case 'task': return `/tasks/${param0}`;
                case 'agent': return `/agents/${param0}`;
                case 'feeling': return `/feelings/${param0}/${param1}`;
                case 'error': return `/errors/${param0}`;
                case 'training': return `/training/${param0}`;
                default: return `/${command.target}`;
            }
        case 'delete':
            switch (command.target) {
                case 'task': return `/tasks/${param0}`;
                case 'agent': return `/agents/${param0}`;
                case 'feeling': return `/feelings/${param0}`;
                case 'error': return `/errors/${param0}`;
                case 'training': return `/training/${param0}`;
                default: return `/${command.target}`;
            }
        case 'search':
            switch (command.target) {
                case 'tasks': return `/tasks/search?q=${param0}`;
                case 'errors': return `/errors/search?q=${param0}`;
                default: return `/${command.target}`;
            }
        case 'run':
            switch (command.target) {
                case 'init': return '/init';
                case 'agent': return `/chat/agent/${param0}`;
                case 'training_export': return '/training/export';
                case 'chat': return '/chat/process';
                case 'time_start': return `/time/start/${param0}`;
                case 'time_stop': return `/time/stop/${param0}`;
                case 'queue_start': return `/queue/start/${param0}`;
                case 'queue_end': return `/queue/end/${param0}`;
                case 'api_register': return '/api/register';
                case 'api_check': return '/api/check';
                default: return `/${command.target}`;
            }
        default:
            return `/${command.target}`;
    }
}

function buildRequestBody(command: TdzCommand): any {
    const getOption = (key: string, defaultValue = ''): string => {
        return command.options.get(key) || defaultValue;
    };
    
    const getArrayOption = (key: string): string[] => {
        const value = command.options.get(key);
        return value ? value.split(',').map(s => s.trim()).filter(s => s.length > 0) : [];
    };
    
    const getNumberOption = (key: string, defaultValue: number): number => {
        const value = command.options.get(key);
        if (value === undefined) return defaultValue;
        const num = parseFloat(value);
        return isNaN(num) ? defaultValue : num;
    };
    
    switch (command.target) {
        case 'task':
            return {
                action: getOption('action'),
                time: getOption('time'),
                priority: getOption('priority'),
                project: getOption('project'),
                status: getOption('status'),
                assignee: command.options.get('assignee'),
                tags: getArrayOption('tags')
            };
        case 'memory':
            return {
                moment: getOption('moment'),
                meaning: getOption('meaning'),
                reason: getOption('reason'),
                importance: getOption('importance'),
                term: getOption('term'),
                memory_type: getOption('memory_type'),
                emotion: command.options.get('emotion')
            };
        case 'idea':
            return {
                idea: getOption('idea'),
                share: getOption('share'),
                importance: getOption('importance')
            };
        case 'agent':
            return {
                name: getOption('name'),
                description: getOption('description'),
                capabilities: getArrayOption('capabilities')
            };
        case 'chunk':
            return {
                id: getOption('id'),
                level: getOption('level'),
                description: getOption('description'),
                dependencies: getArrayOption('dependencies'),
                code: getOption('code')
            };
        case 'project':
            return {
                name: getOption('name'),
                description: getOption('description'),
                status: getOption('status')
            };
        case 'feeling':
            return {
                emotion: getOption('emotion'),
                intensity: getNumberOption('intensity', 5),
                description: getOption('description'),
                context: command.options.get('context'),
                tags: getArrayOption('tags')
            };
        case 'training':
            return {
                data_type: getOption('data_type'),
                prompt: getOption('prompt'),
                completion: getOption('completion'),
                source: getOption('source'),
                context: command.options.get('context'),
                tags: getArrayOption('tags'),
                quality_score: getNumberOption('quality_score', 0)
            };
        default:
            return null;
    }
}

function buildRunBody(command: TdzCommand): any {
    const getOption = (key: string, defaultValue = ''): string => {
        return command.options.get(key) || defaultValue;
    };
    
    switch (command.target) {
        case 'agent':
        case 'chat':
            return {
                message: getOption('message'),
                context: command.options.get('context')
            };
        case 'queue_start':
        case 'queue_end':
            return null;
        case 'api_check':
            return {
                public_key: getOption('public_key'),
                private_key: command.options.get('private_key')
            };
        default:
            return null;
    }
}

export async function processTdzCommands(
    text: string,
    baseUrl: string,
    apiKey?: string
): Promise<any[]> {
    const commands = parseTdzCommand(text);
    const results: any[] = [];
    
    for (const command of commands) {
        const result = await executeTdzCommand(command, baseUrl, apiKey);
        results.push(result);
    }
    
    return results;
}
